Support Item prop style. Provide prop selected to itemRenderer

morulus opened the pull request:
**Issue Number**

Without issue

**Overview of PR**

Allow to specify custom style for the item element. For example, it can be a different background color for each item (driven by some business logic, etc). Strange that it wasn't implemented before. it is so useful feature.

As well as, provide prop `selected` to the itemRenderer. I have a few cases when I need to know inside the item content does it selected or no. For example, I prefer to render some advanced information inside selected items. Or some way react to this prop in the child component lifecycle - open a modal, or send a request at last.

mcMickJuice commented:
Thanks for opening this PR.

Can please open up an issue describing this PR and provide detail about how the current API doesn't give you what you need?

yftachman commented:
Can't you achieve this with the itemRenderer?

morulus commented:
No, itemRenderer gives you access to the inner content of an item, but not to an item itself.

yftachman commented:
Oh, I see what you are saying.
Yeah this feature is needed, in the mean time you can try overriding the item class, might help you out :) Don't forget to use !important

morulus commented:
The class is not suitable for me, because I'm in the situation when item background color is formed dynamically by the server. I can not prepare classes for every possible color :)

mcMickJuice commented:
@morulus can you please open an issue for this PR? No PRs will get merged until there's discussion around its need.

Not saying this won't be added but it overlaps with another issue around refactoring itemRenderer.

morulus commented:
The `style` issue https://github.com/namespace-ee/react-calendar-timeline/issues/347
The `selected` issue  https://github.com/namespace-ee/react-calendar-timeline/issues/348
Moreover, I found the related issue https://github.com/namespace-ee/react-calendar-timeline/issues/244

morulus commented:
I want to consult with you. 
Here is a chunk of code, implements style support:

```js
const style = {
      left: `${dimensions.left}px`,
      top: `${dimensions.top}px`,
      width: `${dimensions.width}px`,
      height: `${dimensions.height}px`,
      lineHeight: `${dimensions.height}px`,
      ...this.props.item.style
    }
```

As you can see, It allows overwriting built-in properties. Probably, it is bad, because this can lead to anomalies. And it would be better to do it in this way:

```js
const style = {
      ...this.props.item.style
      left: `${dimensions.left}px`,
      top: `${dimensions.top}px`,
      width: `${dimensions.width}px`,
      height: `${dimensions.height}px`,
      lineHeight: `${dimensions.height}px`,
    }
```

Thus, the built-in properties will be kept. How do you think?

Ilaiwi commented:
All the usecases for this will be supported when #308 gets merged but  until then this will be useful.

mcMickJuice commented:
Agree with @Ilaiwi .  I might merge this PR but these use cases will be achievable when #308 is merged.  I might prioritize that PR though as it's the direction we want to go anyways.

morulus commented:
Ok. I've updated pr according to https://github.com/namespace-ee/react-calendar-timeline/pull/345#issuecomment-400348465. Changed the order of style in the spread. https://github.com/namespace-ee/react-calendar-timeline/pull/345/commits/caf2cdaa4eee2cb2b7ab7710bcefab463a501ee7

mcMickJuice commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-calendar-timeline",
-  "version": "0.17.0",
+  "version": "0.17.1",
```
For future PRs, don't worry about updating version here. I'll update that as a part of a release

morulus replied:
Oh, I did it, really? Sorry, just a habit)

mcMickJuice commented:
re: https://github.com/namespace-ee/react-calendar-timeline/pull/345#issuecomment-400348465

Yes absolutely, these styles should be overridden by any styles that the timeline is applying (i.e. left, top etc)
